Adding a privacy section in account settings behind feature flag `me/privacy`

bperson opened the pull request:
Introduces a new section in the user settings called `privacy` with a placeholder text for now. It's only available behind the `me/privacy` feature flag, enabled only in `development.json` for now.

<img width="494" alt="screen shot 2018-02-21 at 14 47 16" src="https://user-images.githubusercontent.com/1145270/36483376-37f67c52-1716-11e8-88b9-a9a24849e909.png">

Open questions:

- The `visible` GridIcon might not be the best choice, ideas welcomed.
- I tried to trim as much as possible the controller / main component for now ( compared to other part of `/me` ), but it's still using the `formBase` mixin which might not be ideal?
- It introduces a second `DocumentHead` in the react tree since there is one coming from the default layout. I don't think it's an issue as other places work fine with 2 `DocumentHead`s ( like the Reader ) If we accept this way of doing it, I'll create another PR to remove the `setTitle` dispatched via the context in other controllers under `/me`.

rodrigoi commented on client/me/sidebar/index.jsx:
```diff
@@ -166,6 +167,15 @@ class MeSidebar extends React.Component {
 							preloadSectionName="security"
 						/>
 
+						<SidebarItem
```
this is missing the feature flag enforcing.

gwwar commented:
To test with the feature flag disabled:
```
DISABLE_FEATURES=me/privacy npm start
```

bperson commented:
> DISABLE_FEATURES=me/privacy npm start

TIL, much easier than manually editing `development.json` :D 

I added a check to hide the `SidebarItem` if the flag is disabled. I thought maybe about redirecting folks to `/me` if they come directly to `/me/privacy` while the flag is disabled but maybe it's going a bit too far?

rodrigoi commented:
how complicated things will get for this section? because you can remove `client/me/privacy/index.js` and `client/me/privacy/controller.js` and add a couple lines of code to `client/me/index.js` and `client/me/controller.js`. It will make a smaller diff 😉

bperson commented:
no idea, I'm still not entirely sure about everything that might end up going in there. I'd tend to stay on the cautious side and keep it as close as possible to the rest of the `/me` ecosystem though.

gwwar commented:
@bperson do you mind rebasing? I'm see a stray error, which I believe is unrelated
```
Wrote 8 CSS files to /Users/kerryliu/checkouts/wp-calypso/public/sections
Failed to process scss /Users/kerryliu/checkouts/wp-calypso/client/extensions/woocommerce { Error: File to import not found or unreadable: components/d3/style.
Parent style sheet: /Users/kerryliu/checkouts/wp-calypso/client/extensions/woocommerce/style.scss
    at options.error (/Users/kerryliu/checkouts/wp-calypso/node_modules/node-sass/lib/index.js:291:26)
  status: 1,
  file: '/Users/kerryliu/checkouts/wp-calypso/client/extensions/woocommerce/style.scss',
  line: 35,
  column: 2,
  message: 'File to import not found or unreadable: components/d3/style.\nParent style sheet: /Users/kerryliu/checkouts/wp-calypso/client/extensions/woocommerce/style.scss',
```

gwwar commented on client/me/privacy/controller.js:
```diff
@@ -0,0 +1,19 @@
+/**
+ * Internal dependencies
+ */
+import userSettings from 'lib/user-settings';
```
Huh, I suppose this was never ported to the data-layer? Note that if we need to update a user's preference vs settings, preferences are available from the global redux store.

gwwar commented on client/me/privacy/main.jsx:
```diff
@@ -0,0 +1,41 @@
+const Privacy = createReactClass( {
+	displayName: 'Privacy',
+
+	mixins: [ formBase ],
```
What does formBase do? We can try porting functionality to HoC if we really need it.

gwwar replied:
Ah so if we look at the readme, there are a few options here. There's both a [HoC](https://github.com/Automattic/wp-calypso/tree/aafc155b7c3d67d054b29fc32eccc07655f0a2e2/client/lib/protect-form#usage-hoc) available in addition to a [declarative component](https://github.com/Automattic/wp-calypso/tree/aafc155b7c3d67d054b29fc32eccc07655f0a2e2/client/lib/protect-form#usage-declarative).

Let's remove this for now and add it back when we actually have a form or toggle to protect changes for.

gwwar commented on client/me/privacy/main.jsx:
```diff
@@ -0,0 +1,41 @@
+// NOTE: disabling it since we need to use the `formBase` mixin
+// eslint-disable-next-line react/prefer-es6-class
+const Privacy = createReactClass( {
+	displayName: 'Privacy',
```
Let's make this a standard React component. 

Also, setting the `displayName` isn't necessary here, since we use the `transform-react-display-name` babel transform. (One exception is for example components in DevDocs since we'd like the displayName to not be shortened by the minifier.)

gwwar commented:
The me section looks rather large, so I don't mind letting privacy have it's own controller and such. Main thing blocking right now is that that Privacy is using createClass, and also is using a mixin. Let's update that to a plain React component, and this should be good to 🚢

bperson commented:
> Privacy is using createClass, and also is using a mixin. Let's update that to a plain React component.

For now, I've removed the `formBase` mixin ( and the `protectForm` HoC ) since they are not strictly needed in this PR. As a result, I was able to migrate it to a plain React component.

See #22693 for an attempt at HoC-ing `formBase` in preparation for future PRs in the privacy section.

> I can visit http://calypso.localhost:3000/me/privacy directly with the config flag disabled. We should be redirected when this occurs.

Added a redirection to `/me` if the flag is disabled. I'm curious: how far do we usually go? I've another PR incoming for the opt-out introducing a new Redux action, should the action be gated by the feature flag as well so that someone messing in the JS console cannot trigger it in prod?

gwwar commented:
> Added a redirection to /me if the flag is disabled. I'm curious: how far do we usually go?

General rule of thumb would be, can a user trigger any behavior from normal browsing? So adding new actions is fine, but I'd expect we wouldn't dispatch them outside of a feature check.

gwwar commented on client/me/privacy/controller.js:
```diff
@@ -0,0 +1,23 @@
+
+export default function privacyController( context, next ) {
+	if ( ! config.isEnabled( 'me/privacy' ) ) {
+		page.redirect( '/me' );
```
Stylistically we prefer early returns if possible. This reduces the need for if/else blocks.

So this may look like:
```js
if ( ! config.isEnabled( 'me/privacy' ) ) {
    return page.redirect( '/me' );
}
context.primary = React.createElement( PrivacyComponent );
next();
```

kwight commented:
Everything works as described for me 👍 

> Privacy is using createClass, and also is using a mixin. Let's update that to a plain React component.

@gwwar I've never worked with removing mixins before – since they're by purpose reusable, could we add the HOC version to `/client/components`?

gwwar commented:
@kwight see: https://github.com/Automattic/wp-calypso/pull/22648#discussion_r169806888
